### ISSUE TITLE
`unicorn/no-unnecessary-array-flat-map` changes type

### ISSUE THREAD

**Issue by KieranP** [sindresorhus/eslint-plugin-unicorn#3496]
**Before rule:**

```ts
const types = new Set(
  records.flatMap((r) =>
    newIds.includes(r._id) && r.type
      ? [r.type]
      : []
  )
)
```

TS Type: `Set<RecordType>`

**After rule:**

```ts
const types = new Set(
  records
    .filter((r) => newIds.includes(r._id) && r.type)
    .map((r) => r.type)
)
```

TS Type: `Set<RecordType | undefined>`

**Cause:**

TypeScript does a better job of determining the return type for flatMap than for filter/map. So this new rule is suggesting a change that actually changes resulting types. Granted, this is more an issue for TypeScript, since they should (possible) be able to infer this properly, but since this is enabled by default and auto fixed, it could lead to some type breakages, as it has on our end.
